DOC: Refine SVD documentation

tovrstra opened the pull request:
Fixes #7828

This is an attempt to fix some shortcomings for the SVD docstring:

- Dimensions of inputs and outputs were not completely correct.
- Name of the third return value was confusing. Changed `v` by `vh`. See #7828 for details.
- Add link to broadcasting for linalg functions.

tovrstra committed:
Refine SVD documentation

bashtage commented on numpy/linalg/linalg.py:
```diff
@@ -1300,19 +1300,21 @@ def svd(a, full_matrices=True, compute_uv=True):
 
-    Factors the matrix `a` as ``u * np.diag(s) * v``, where `u` and `v`
-    are unitary and `s` is a 1-d array of `a`'s singular values.
+    Factors the matrix `a` as ``u * np.diag(s) * vh``. When `a` is a 2D array,
```
Should `*` by `@` since the latter is now meaningful?

tovrstra replied:
Good for me. Do you know if this is common or desirable in the numpy docs?

bashtage commented:
I don't know, so should see what committers think.

eric-wieser commented on numpy/linalg/linalg.py:
```diff
@@ -1300,19 +1300,21 @@ def svd(a, full_matrices=True, compute_uv=True):
-        If True (default), `u` and `v` have the shapes (`M`, `M`) and
-        (`N`, `N`), respectively.  Otherwise, the shapes are (`M`, `K`)
-        and (`K`, `N`), respectively, where `K` = min(`M`, `N`).
+        If True (default), `u` and `v` have the shapes (..., `M`, `M`) and
```
Should this also be vh?

tovrstra replied:
Indeed. I'll double check for similar mistakes

tovrstra committed:
Improve SVD docstring further

- Use @ instead of * for matrix multiplication in docstring
- One more v had to be renamed to vh
- Paragraph with algebraic details contained a lot of redundant
  statements and is now simplified.
- np.multi_dot in example, making it easier to read.
- One more empty line after SVD function (PEP8)

tovrstra commented:
I've fixed all comments given so far and also made a few additional improvements. (See commit message.)

mhvk commented:
I have been confused by the documentation myself; good to see it clarified! Looks all OK to me.

charris commented on numpy/linalg/linalg.py:
```diff
@@ -1300,19 +1300,21 @@ def svd(a, full_matrices=True, compute_uv=True):
 
-    Factors the matrix `a` as ``u * np.diag(s) * v``, where `u` and `v`
-    are unitary and `s` is a 1-d array of `a`'s singular values.
+    Factors the matrix `a` as ``u @ np.diag(s) @ vh``. When `a` is a 2D array,
```
Hmm, `diag` is a nasty function. Could write this as ` (u * s) @ vh` (non- stacked), or `(u * s[..., None, :]) @ vh` for the stacked case. I don't know if that is any clearer, though.

charris replied:
Which is to say, that those two forms are greatly preferred for actual code, but might not be the best from a pedagogical standpoint.

charris replied:
Might put those alternatives in the `Notes` section.

eric-wieser replied:
`diag` fails in confusing ways for the stacked case, which is an argument against using it.

These are correct too, right?
* `u @ (s[...,None] * vh)`
* `u @ (s[...,None]*np.eye(s.shape[-1]) @ vh`

charris replied:
Those examples also look correct. I prefer the original as they are more likely to be cache friendly, i.e., contiguous access.

charris replied:
Although the right hand versions look cleaner, so might be better for newbies.

eric-wieser replied:
Could also introduce (in another PR) a `np.linalg.diag` with signature `(...,N) -> (...,N,N)` that behaves correctly on stacks of vectors, and doesn't have the weird overloading that `np.diag` has

tovrstra replied:
What about first explaining the non-stacked case, both with `diag` and the more efficient approach?

Further down, the stacked case can be discussed. Because that is never really going to be trivial in terms of python code, I would directly go for `u @ (s[...,None] * vh)` and skip the `eye` method. (I also prefer the right-hand version because it is a little shorter.)

eric-wieser replied:
I think keeping a special case for 2d using `diag` is probably the right thing to do.

> as they are more likely to be cache friendly, i.e., contiguous access.

Are you sure about that? Are the `vh` and `u` matrices fortran-order?

charris replied:
They all seem to be C contiguous.

tovrstra replied:
The comments on the contiguity and caching are not completely obvious to me. Both `s[..., None, :]` and `s[..., None]` are C contiguous, which seems fine, so I guess the problem is in the broadcasting of the element-wise multiplication.

Say we have a matrix `a` with shape `(5, 4, 3)`:

- Case `(u * s[..., None, :])`. I guess `u` is traversed contiguously, so 4 passes through `s` are needed when computing this product.

- Case `(s[..., None] * vh)`. I guess `vh` is traversed contiguously, so only 1 pass through `s` is needed when computing this product.

Isn't the second case better for caching? (I'm not sure how Numpy loops through the arrays internally.)

Are  such performance issues relevant compared to the cost of SVD?

charris replied:
Yeah, it isn't clear how the actions are actually ordered, but one way would be to skip down the columns of `vh` multiplying by the elements of `s`, the skipping breaking the cache, while in the first case the elements would all be accessed sequentially in memory, which would be cache friendly. But as you say, it could be that the rows of `vh` might be multiplied by a single element, which would also be cache friendly. In practice, it looks like the second is a bit better
```
In [1]: u = array([eye(3)]*1000000)

In [2]: s = array([[1,2,3]]*1000000)

In [3]: %timeit s[...,None]*u
10 loops, best of 3: 30.2 ms per loop

In [4]: %timeit u*s[...,None,:]
10 loops, best of 3: 32.8 ms per loop
```
Talking about cache is probably out of scope for a docstring, so I would go with the example that looks best. Both are more efficient than converting the singular values to matrices ...

tovrstra replied:
Thanks for explaining. I'll further edit the docs in line with the discussions here.

tovrstra committed:
Further improvements to the SVD docstring

tovrstra commented:
I tried to incorporate most of the suggestions made so far. I've also made sure the doctests pass (for python < 3.5). I've made a few additional changes to make the docstring more precise and a little easier to understand. Suggestions are obviously still welcome!

charris commented on numpy/linalg/linalg.py:
```diff
@@ -1336,46 +1345,75 @@ def svd(a, full_matrices=True, compute_uv=True):
+
+    SVD is usually described for the factorization of a 2D matrix :math:`A`.
+    The higher-dimensional case will be discussed below. In the 2D case, SVD is
+    written as :math:`A = U S V^H``, where :math:`A=` ``a``, :math:`U=` ``u``,
```
B^H`` ? Too many backticks?

charris replied:
Maybe just put the whole bit in latex, ":math:`A=a` ... :math:`B^H=vh`"?

eric-wieser replied:
Or ``:math:`A = \mathtt{a}` ``, if that's supported

eric-wieser replied:
Or ``:math:`A = \mathtt{a}` ``, if that's supported

tovrstra replied:
I've tried `\mathtt` and it works but the result is not very intuitive. It seems to be common in numpy docstrings to write variable names in italics, with single back ticks, while `\mathtt` does the opposite in Latex. Single back ticks looks more consistent.

charris commented on numpy/linalg/linalg.py:
```diff
@@ -1336,46 +1345,75 @@ def svd(a, full_matrices=True, compute_uv=True):
+    SVD is usually described for the factorization of a 2D matrix :math:`A`.
+    The higher-dimensional case will be discussed below. In the 2D case, SVD is
+    written as :math:`A = U S V^H``, where :math:`A=` ``a``, :math:`U=` ``u``,
+    :math:`S=` ``np.diag(s)`` and :math:`V^H=` ``vh``. `s` is then a 1D array
```
Maybe "The singular values `s` ..." so that the sentence begins with a capital.

tovrstra replied:
ok

charris commented on numpy/linalg/linalg.py:
```diff
@@ -1336,46 +1345,75 @@ def svd(a, full_matrices=True, compute_uv=True):
+    The higher-dimensional case will be discussed below. In the 2D case, SVD is
+    written as :math:`A = U S V^H``, where :math:`A=` ``a``, :math:`U=` ``u``,
+    :math:`S=` ``np.diag(s)`` and :math:`V^H=` ``vh``. `s` is then a 1D array
+    with the singular values of `a` and `u` and `vh` are unitary: the rows of `vh`
```
I'd make a new sentence rather than use ":", which doesn't seem right in any case. Or maybe ", where the rows ..."

tovrstra replied:
ok

tovrstra committed:
Fix few more small things in SVD docs

charris commented on numpy/linalg/linalg.py:
```diff
@@ -1336,46 +1345,75 @@ def svd(a, full_matrices=True, compute_uv=True):
+    the singular values of `a` and `u` and `vh` are unitary. The rows of `vh`
+    are the eigenvectors of :math:`A^H A` and the columns of `u` are the
+    eigenvectors of :math:`A A^H`. For row `i` in `vh` and column `i` in `u`,
+    the corresponding eigenvalue is `s[i]**2`.
```
Double backticks. The single backticks are for links or content following a directive.

charris replied:
In other words, do not use single backticks for "i" either, as it is not something that can be linked to, but is a bit of code. Maybe say "i-th" column instead, which has the advantage of being english.

charris replied:
I checked to make sure, but apart from directives, single backticks are used for generating links.

tovrstra replied:
Thank for clarifying this. It did not appear to me what to conventions were so far. Let me try to get this right.

tovrstra committed:
fix backticks and versionadded in SVD docs

tovrstra commented:
Full disclosure: I'm developing a backtick fetish. :)

Another thing I noticed is that the `.. versionadded:` directive was out of place. It only applies to the broadcasting feature. I've moved it to the bottom to make sure that one would not get the impression that SVD as a whole was added in numpy 1.8.0.

charris commented on numpy/linalg/linalg.py:
```diff
@@ -1300,31 +1300,40 @@ def svd(a, full_matrices=True, compute_uv=True):
-        If True (default), `u` and `v` have the shapes (`M`, `M`) and
-        (`N`, `N`), respectively.  Otherwise, the shapes are (`M`, `K`)
-        and (`K`, `N`), respectively, where `K` = min(`M`, `N`).
+        If True (default), `u` and `vh` have the shapes (..., `M`, `M`) and
```
I've gone ahead and generated to docs to see what this looks like. The shapes can break across lines, so I think it would be better to enclose them in double backticks: \`\` (..., M, M)\`\`, etc.  Same below when defining `K`.

tovrstra replied:
ok

charris commented on numpy/linalg/linalg.py:
```diff
@@ -1334,48 +1343,77 @@ def svd(a, full_matrices=True, compute_uv=True):
 
-    The SVD is commonly written as ``a = U S V.H``.  The `v` returned
-    by this function is ``V.H`` and ``u = U``.
+    SVD is usually described for the factorization of a 2D matrix :math:`A`.
```
Here I tried
```
    SVD is usually described for the factorization of a 2D matrix :math:`A`.
    The higher-dimensional case will be discussed below. In the 2D case, SVD is
    written as :math:`A = U S V^H`, where :math:`A = a`, :math:`U= u`,
    :math:`S= \mathtt{np.diag}(s)` and :math:`V^H = vh`. The 1D array `s` contains
    the singular values of `a` and `u` and `vh` are unitary. The rows of `vh`
    are the eigenvectors of :math:`A^H A` and the columns of `u` are the
    eigenvectors of :math:`A A^H`. In both cases the corresponding (possibly
    non-zero) eigenvalues are given by ``s**2``.
```
which I think looks a touch better and also avoids the potential to break across lines.

charris replied:
Note that some of the lines are also >= 80 characters, they should be < 80.

tovrstra replied:
ok

charris commented on numpy/linalg/linalg.py:
```diff
@@ -1334,48 +1343,77 @@ def svd(a, full_matrices=True, compute_uv=True):
-    of `u` are the eigenvectors of ``a a.H``.  For row ``i`` in
-    `v` and column ``i`` in `u`, the corresponding eigenvalue is
-    ``s[i]**2``.
+    .. versionadded:: 1.8.0
```
I'd move the `.. versionadded::` up to the top of the notes with a one line explanation, see other functions where that is done. The expanded explanation can be left here, maybe beginning `If a has more than two dimensions, then broadcasting rules apply as explained in ....`

tovrstra replied:
sounds good too.

tovrstra replied:
One more question. I've seen two styles to do `.. versionadded` in linalg.py:

```
    .. versionchanged:: 1.14
       Can now operate on stacks of matrices
```

and

```
    .. versionadded:: 1.8.0

    Broadcasting rules apply, see the `numpy.linalg` documentation for
    details.
```

The former seems clearer to me but is less common. It renders approximately as:

Changed in version 1.14: Can now operate on stacks of matrices

I.e. everything on one line, so it is clear that following paragraphs are not related to changes in a given version. Would you agree this is better?

charris replied:
Hard to pick between the two, but the first does seem a hair better. Note that the indented message also works for `.. versionadded::`

tovrstra replied:
Now I get it. Just to be sure, I checked also the documentation of the `os` module of Python, which contains plenty of examples and I also checked the Sphinx docs to make sure I get it right.

A lot of docstrings in linalg.py, which mostly use `versionadded` where it should in most cases be `versionchanged`, are missing the indentation. This confused when I was looking at other examples. I'd rather do it right here.

charris commented:
I can put this up with the suggested mods up as a new PR if you wish.

tovrstra commented:
I'm fine either way. I have time in a few hours to take care of all points.

charris commented:
If you have the time, go for it.

tovrstra committed:
Fixed few more issues in SVD docs:

- position and usage of versionchanged directive
- more backticks

charris committed:
DOC: Fix Python 3.6 illegal escape sequence `\m`

charris commented:
Looks good. I fixed the illegal `\m` that caused the failing tests. The link to `numpy.linalg` is failing, but that is true everywhere, so out of scope for this PR. I'll merge this when the tests finish. Thanks @tovrstra .

tovrstra commented:
You're welcome!
